Cellular: Basic driver for Quectel EC25/EC21

hasnainvirk opened the pull request:
### Description

Provides basic power up / power down sequences for Quectel EC2X series
modems. Can be used in PPP mode. For using the on-board IP stack, we
will need to add and implement classes that provide context.

Driver constructor takes power and reset control pins along with the
FileHandle. A default construction is provided which can be chosen by
the application in its mbed_app.json. Otherwise the user is free to
construct as per demand.


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

Commit:
Basic driver for Quectel EC25/EC21

Provides basic power up / power down sequences for Quectel EC2X series
modems. Can be used in PPP mode. For using the on-board IP stack, we
will need to add and implement classes that provide context.

Driver constructor takes power and reset control pins along with the
FileHandle. A default construction is provided which can be chosen by
the application in its mbed_app.json. Otherwise the user is free to
construct as per demand.

Commit:
Removing left over tracing

A test trace was left lurking around. Removed in this commit.

AriParkkila commented on features/cellular/framework/targets/QUECTEL/EC2X/QUECTEL_EC2X.cpp:
```diff
@@ -0,0 +1,132 @@
+    AT_CellularNetwork::RegistrationModeLAC,    // C_EREG
+    AT_CellularNetwork::RegistrationModeLAC,    // C_GREG
+    AT_CellularNetwork::RegistrationModeEnable,    // C_REG
+    1,  // AT_CGSN_WITH_TYPE
```
Should be 0?

hasnainvirk replied:
Why ? AT+CGSN command is supported by the modem. Am I missing something here ?

AriParkkila replied:
Yes, it supports the command but without <snt> parameter, see 3GPP TS 27.0.0.7 chapter 5.4.

AriParkkila commented on features/cellular/framework/targets/QUECTEL/EC2X/QUECTEL_EC2X.cpp:
```diff
@@ -0,0 +1,132 @@
+static const intptr_t cellular_properties[AT_CellularBase::PROPERTY_MAX] = {
+    AT_CellularNetwork::RegistrationModeLAC,    // C_EREG
+    AT_CellularNetwork::RegistrationModeLAC,    // C_GREG
+    AT_CellularNetwork::RegistrationModeEnable,    // C_REG
```
Could be RegistrationModeLAC?

hasnainvirk replied:
Yeah, let me change that.

AriParkkila commented on features/cellular/framework/targets/QUECTEL/EC2X/QUECTEL_EC2X.cpp:
```diff
@@ -0,0 +1,132 @@
+
+    _at->set_at_timeout(5000);
+    _at->resp_start();
+    _at->set_stop_tag("RDY");
```
Do we get "RDY" also if modem has already been powered on?

hasnainvirk replied:
Modem is already powered on here. In soft_power_on() we are just resetting the modem just in case of a jittery behaviour. And we will get RDY.

Commit:
CREG can handle lac/ci/lat

AT+CREG command can also provide information regarding lac/ci/lat etc so
we should set the mode correctly in cellular properties structure.

Commit:
AT+CGSN with EC2X does not take parameter

Standard 3GPP TS27.007 states that the AT+CGSN command can optionally
have type parameter that determines the output. However Quiectel EC2X
series doesn't accept parameters. We need to reflect that in the
cellular properties setup.

kivaisan commented on features/cellular/framework/targets/QUECTEL/EC2X/QUECTEL_EC2X.h:
```diff
@@ -0,0 +1,50 @@
+#ifndef QUECTEL_EC2X_H
+#define QUECTEL_EC2X_H
+
+#ifdef TARGET_FF_ARDUINO
```
These should be defined in mbed_lib.json as target_overrides.

hasnainvirk replied:
Would argue to remove these. Not sure of their exact purpose. While defining pins in your mbed_app.json you can alwyas spell out an arduino pin name convention. It's no different than providing any other pin name, e.g., PE15 or D1, it's just a convention.

kivaisan replied:
If not sure about purpose, then remove completely. It is really annoying and confusing that some target specific settings are in code.

0xc0170 replied:
> Would argue to remove these.

Why would we want to leave this in - setting tx/rx to Dx pins?

AnttiKauppila replied:
Compilation error is enough and correct way to handle this

kivaisan commented on features/cellular/framework/targets/QUECTEL/EC2X/QUECTEL_EC2X.cpp:
```diff
@@ -0,0 +1,132 @@
+nsapi_error_t QUECTEL_EC2X::hard_power_off()
+
+{
+    if (_pwr_key.is_connected()) {
```
`hard_power_on()` and `hard_power_off()` are very much the same duplicate code (only small delay difference), so how about just having the implementation in `hard_power_off()` and calling that from `hard_power_on()` as well?

kivaisan commented on features/cellular/framework/targets/QUECTEL/EC2X/QUECTEL_EC2X.cpp:
```diff
@@ -0,0 +1,132 @@
+    return NSAPI_ERROR_OK;
+}
+
+nsapi_error_t QUECTEL_EC2X::soft_power_off()
```
This also could reuse just one mehtod which toggles power pin.

Commit:
Tidying up the boiler plate code

A little tidying up was needed as there was some boiler plate code that
could be shoved into a corner and reused.

Commit:
Adding pin polarity and changing the constructor

To make this driver more useful, we needed to do some changes. The
wirings for the modem can differ on the board so we also need to be
flexible in our approach.
It is now mandatory to provide the power pin and pin polarity in the
constructor alongwith the file handle. Reset pin is optional.

AriParkkila reviewed (approved):
Is it intentionally assumed that pwr and rst pin polarity are the same?

Commit:
Marking TX and RX pins for UARTSerial as NC

For default construction of the driver, it would be necessary to define
the pins. However as we don't know the correct pins, until/unless user
sets them up for us. So it makes sense to mark the pins as NC. It would
be easier to debug if the pins are set to NC if user forgot to set them
up.
